src/types/transaction.zig

```diff
@@ -1,139 +1,211 @@
+
+    return tx;
+}
 
-    /// Calculate the virtual size of the transaction
-    pub fn virtual_size(self: *const Transaction) usize {
-        // This is a simplified size calculation. In a real implementation,
-        // you would need to account for segregated witness data if present.
-        var size: usize = 8; // Version (4 bytes) + LockTime (4 bytes)
-        size += self.inputs.items.len * 41; // Simplified input size
-        size += self.outputs.items.len * 33; // Simplified output size
-        return size;
+/// Deserialize bytes into Self
+pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !Self {
+    var fbs = std.io.fixedBufferStream(bytes);
+    return try Self.deserializeReader(allocator, fbs.reader());
+}
+
+/// Initialize a new transaction
+pub fn init(allocator: std.mem.Allocator) !Self {
+    return Self{
+        .version = 1,
+        .inputs = &[_]Input{},
+        .outputs = &[_]Output{},
+        .lock_time = 0,
+        .allocator = allocator,
+    };
+}
+
+/// Deinitialize the transaction
+pub fn deinit(self: *Self) void {
+    for (self.inputs) |*input| {
+        input.deinit();
+    }
+    for (self.outputs) |*output| {
+        output.deinit();
     }
-};
+    self.allocator.free(self.inputs);
+    self.allocator.free(self.outputs);
+}
+
+/// Add an input to the transaction
+pub fn addInput(self: *Self, previous_outpoint: OutPoint) !void {
+    const script_sig = try Script.init(self.allocator);
+    const new_capacity = self.inputs.len + 1;
+    var new_inputs = try self.allocator.realloc(self.inputs, new_capacity);
+
+    self.inputs = new_inputs;
+
+    new_inputs[self.inputs.len - 1] = Input{
+        .previous_outpoint = previous_outpoint,
+        .script_sig = script_sig,
+        .sequence = 0xffffffff,
+    };
+}
+
+/// Add an output to the transaction
+pub fn addOutput(self: *Self, value: i64, script_pubkey: Script) !void {
+    var new_script = try Script.init(self.allocator);
+    try new_script.push(script_pubkey.bytes);
+
+    const new_capacity = self.outputs.len + 1;
+    var new_outputs = try self.allocator.realloc(self.outputs, new_capacity);
+
+    self.outputs = new_outputs;
+
+    new_outputs[self.outputs.len - 1] = Output{
+        .value = value,
+        .script_pubkey = new_script,
+    };
+}
+
+/// Calculate the transaction hash
+pub fn hash(self: *const Self) Hash {
+    var h: [32]u8 = undefined;
+    std.crypto.hash.sha2.Sha256.hash(@as([]const u8, std.mem.asBytes(&self.version)), &h, .{});
+    return Hash{ .bytes = h };
+}
+
+/// Calculate the virtual size of the transaction
+pub fn virtual_size(self: *const Self) usize {
+    // This is a simplified size calculation. In a real implementation,
+    // you would need to account for segregated witness data if present.
+    var size: usize = 8; // Version (4 bytes) + LockTime (4 bytes)
+    size += self.inputs.len * 41; // Simplified input size
+    size += self.outputs.len * 33; // Simplified output size
```

Comment:
@sizeOf(Input) ?

Reply:
Same

Reply:
this one could be a behaviour change so let's not do it